### ISSUE TITLE
__exit__ should not return anything unless it has handled exceptions

### ISSUE THREAD

**Issue by geokala** [p4r4n0y1ng/pyfmg#2]
Hi,
Currently, __exit__ returns the result of the logout call. This is incorrect, as it results in suppression of  all exceptions generated while using this as a context manager.

Specific line of code: https://github.com/p4r4n0y1ng/pyfmg/blob/master/pyFMG/fortimgr.py#L182

Relevant documentation: https://docs.python.org/2/reference/datamodel.html#object.__exit__

**Comment by p4r4n0y1ng** [comment]
Hey great point! Okay, so......this was written in a manner where I was actually suppressing the exceptions purposefully and needed the logout info - but, the more I look at it, the more I really don't need the logout information, just to know that it did logout.

So I'm thinking:
`def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type is None and exc_val is None and exc_tb is None:
            self.logout()
            return False
        return True`

What say you?

**Comment by mwclark4453** [comment]
It would help to not lose the exception.  I've had several times when I've done something I shouldn't have in python and the context just exits without telling me what happened.

For instance.  When I add this before the "with" context for pyFMG:
thisouttobreak()
I get back the exception:
Traceback (most recent call last):
  File "./policy_fmg2copy.py", line 660, in <module>
    thisouttobreak()
NameError: name 'thisouttobreak' is not defined

but when I move it to below the "with", all that happens is I log out of the FMG that I have just logged into. This makes troubleshooting any significant amount of code very difficult.

**Comment by geokala** [comment]
I'm with mwclark4453- I think this should let exceptions bubble up. If they're expected and not cared about then they can be handled by something higher up the stack.

e.g.
def __exit__(self, exc_type, exc_val, exc_tb):
    self.logout()
    return True

**Comment by p4r4n0y1ng** [comment]
Agreed. Fixing....I think the geokala way is the most pythonic and most clear and concise (which is like saying the same thing twice ;) I know)

**Comment by AaronFoley** [comment]
Hi,

I don't believe this has fixed the issue. Returning a True value causes the exception to be suppressed by a with statement:

> If an exception is supplied, **and the method wishes to suppress the exception (i.e., prevent it from being propagated), it should return a true value**. Otherwise, the exception will be processed normally upon exit from this method.

[https://docs.python.org/2/reference/datamodel.html#object.__exit](url)

I believe it should be defined as:

```
def __exit__(self, exc_type, exc_val, exc_tb):
    self.logout()
```

**Comment by p4r4n0y1ng** [comment]
Yes, I was actually suppressing the exception - it's how we were using it. However, after reading this, I think what everyone is asking for is for the exception to actually BE propogated. I will remove the return true and just logout and let the exception go if that's what everyone wants. Probably better anyway. Let me finish the locking context and then I'll mod this back in that branch. Also, I am going to be adding an exception object class in this and have those all created and handled later as well. Anyway thanks Aaron!

**Comment by geokala** [comment]
Agreed- I was intending to ask for it to be propagated, I must've had a braino in my earlier comment- sorry!

Thanks Aaron for catching that!

**Comment by p4r4n0y1ng** [comment]
Fixed in 0.7.0 - which is also where I added Workspace/Flow capabilities. Thanks!
